src/prove.rs

```diff
@@ -0,0 +1,15 @@
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// FIXME
+///
+/// ```mermaid
+/// flowchart LR
+///   Invocation --> more --> Self --> Candidate --> more2
+///   more[...]
+///   more2[...]
+/// ```
+pub trait TryProve<'a, T> {
+    type Error;
+    type Proven;
+
+    fn try_prove(&'a self, candidate: &'a T) -> Result<&'a Self::Proven, Self::Error>;
+}
```

Comment:
I suspect that this version of the trait won't quite work, or at least I'm confused about how it would.

To write a function that takes an `Invocation` and verifies that some `A: Ability` can be proven, that function will need to take the `proofs: Vec<Cid>` in the `Invocation` and start fetching and parsing them to valid `Delegation`s.
But what type would you use for the abilities in these `Delegation`s?
Likely it'll be the same ability type for all elements in the `Vec`, because it needs to be homogenous.
However, `trait TryProve<'a, T>` allows `self`, the `candidate` and return type `Self::Proven` to be different types.

Grepping all `impl TryProve`s it seems like in most cases you're doing `impl TryProve<X> for X` with `type Proven = X`.
So perhaps it should just be `fn try_prove(&self, candidate: &Self) -> Result<Self, Self::Error>`?
The only exception is the `Crud` ability. In that case it looks like there exist e.g. `impl TryProve<Crud> for CrudRead` with `type Proven = CrudRead`. I'm not sure exactly what you're trying to solve, but my gut reaction would be to simply have an `enum Crud { CrudAll, CrudRead, CrudWrite, ... }`.

Reply:
Ah, it _does not_ take an invocation! It only takes Abilities (command + args). The validation around that just depends on this being implemented for the `T` in `Invocation<T>` / `Delegation<T>` / `Receipt<T>`

Reply:
Hmm. I'm confused. Let me try to get clarification:
1. There will be a `Invocation::<T>::check(&self, ability: impl TryProve<...>, ...) -> Result<...>` function that verifies the given ability is valid in the invocation, right? Or perhaps that's even just `Invocation::<T>::check(&self, ...) -> Result<...>`, and it'll just return the ability in the invocation or sth.
2. The invocation checking code will call the `TryProve` implementations of your abilities, right?